File: backend/services/campaign_chatbot.py
"""
Campaign Chatbot Service
Handles AI-powered campaign creation with conversation flow and smart suggestions
"""

from typing import Dict, Any, List, Optional
from datetime import datetime
import json
import uuid
from services.logger import app_logger
from services.mongo_db import mongo_db
from services.util import get_user_from_token
from services.poster_generator import poster_generator

class CampaignChatbot:
    """AI-powered campaign creation assistant"""
    
    def __init__(self):
        self.conversation_steps = [
            'welcome',
            'campaign_name',
            'campaign_type',
            'target_audience',
            'duration',
            'budget',
            'platforms',
            'goal',
            'visual_theme',
            'special_offers',
            'call_to_action',
            'additional_details',
            'poster_generation',
            'campaign_creation'
        ]
        
        self.campaign_types = {
            'sale': {
                'visual_theme': 'blue_ocean',
                'default_hashtags': ['#Sale', '#SpecialOffer', '#Discount', '#Deals'],
                'call_to_action': 'Shop Now',
                'ad_copy_templates': [
                    "🔥 {name} is ON! Get {special_offers}! {cta} →",
                    "☀️ Hot {name} Deals! {special_offers}. Limited time only!",
                    "🛍️ {name} Shopping Spree! {special_offers}!"
                ]
            },
            'product_launch': {
                'visual_theme': 'modern_tech',
                'default_hashtags': ['#NewProduct', '#Innovation', '#Launch', '#Tech'],
                'call_to_action': 'Learn More',
                'ad_copy_templates': [
                    "🚀 Exciting News! {name} is here. Discover the future today!",
                    "✨ Innovation Unveiled! Be the first to experience {name}.",
                    "🎯 Game Changer Alert! Transform your experience with {name}."
                ]
            },
            'holiday': {
                'visual_theme': 'festive_red',
                'default_hashtags': ['#Holiday', '#Festive', '#Special', '#Celebration'],
                'call_to_action': 'Sign Up',
                'ad_copy_templates': [
                    "🎄 {name} Magic! {special_offers}. {cta} now!",
                    "🎅 Festive {name} Special! {special_offers}. Join us!",
                    "🎁 {name} Cheer! {special_offers}. Don't miss out!"
                ]
            },
            'brand_awareness': {
                'visual_theme': 'elegant_black',
                'default_hashtags': ['#Brand', '#Discover', '#Story', '#Identity'],
                'call_to_action': 'Discover',
                'ad_copy_templates': [
                    "✨ Discover {name} - A story worth sharing.",
                    "🌟 {name} - More than just a brand, it's an experience.",
                    "🎯 {name} - Redefining excellence in every way."
                ]
            },
            'lead_generation': {
                'visual_theme': 'professional_green',
                'default_hashtags': ['#Leads', '#Growth', '#Business', '#Success'],
                'call_to_action': 'Sign Up',
                'ad_copy_templates': [
                    "📈 Grow your business with {name}! {special_offers}. {cta}!",
                    "🎯 {name} - Your path to success starts here. {cta} now!",
                    "💼 Transform your business with {name}. {special_offers} inside!"
                ]
            }
        }
        
        self.target_audiences = {
            'new_customers': {
                'optimal_times': ['12:00 PM', '7:00 PM'],
                'content_focus': 'introductory offers and value proposition'
            },
            'existing_customers': {
                'optimal_times': ['9:00 AM', '6:00 PM'],
                'content_focus': 'loyalty rewards and exclusive deals'
            },
            'all_customers': {
                'optimal_times': ['10:00 AM', '2:00 PM', '8:00 PM'],
                'content_focus': 'broad appeal and general promotions'
            },
            'specific_demographics': {
                'optimal_times': ['11:00 AM', '5:00 PM'],
                'content_focus': 'targeted messaging and personalized content'
            }
        }
        
        self.platform_strategies = {
            'facebook': {
                'content_type': 'carousel posts and videos',
                'optimal_length': 'short to medium',
                'engagement_tips': 'use questions and interactive content'
            },
            'instagram': {
                'content_type': 'visual stories and reels',
                'optimal_length': 'visual-first',
                'engagement_tips': 'use high-quality images and trending audio'
            },
            'linkedin': {
                'content_type': 'professional articles and company updates',
                'optimal_length': 'medium to long',
                'engagement_tips': 'focus on value and professional insights'
            },
            'twitter': {
                'content_type': 'quick updates and threads',
                'optimal_length': 'short and punchy',
                'engagement_tips': 'use hashtags and trending topics'
            }
        }

    def process_user_input(self, step: str, user_input: str, campaign_data: Dict[str, Any]) -> Dict[str, Any]:
        """Process user input at each conversation step"""
        
        try:
            if step == 'campaign_name':
                campaign_data['name'] = user_input.strip()
                return campaign_data
                
            elif step == 'campaign_type':
                campaign_type = self._map_campaign_type(user_input)
                campaign_data['type'] = campaign_type
                return campaign_data
                
            elif step == 'target_audience':
                audience = self._map_target_audience(user_input)
                campaign_data['target_audience'] = audience
                return campaign_data
                
            elif step == 'duration':
                duration = self._map_duration(user_input)
                campaign_data['duration_days'] = duration
                return campaign_data
                
            elif step == 'budget':
                campaign_data['budget_range'] = user_input.strip()
                return campaign_data
                
            elif step == 'platforms':
                platforms = self._map_platforms(user_input)
                campaign_data['platforms'] = platforms
                return campaign_data
                
            elif step == 'goal':
                goal = self._map_goal(user_input)
                campaign_data['goal'] = goal
                return campaign_data
                
            elif step == 'visual_theme':
                theme = user_input.strip().lower()
                valid_themes = ['blue_ocean', 'modern_tech', 'festive_red', 'nature_green', 
                               'elegant_purple', 'sunset_orange', 'elegant_black', 'professional_green']
                
                if theme in valid_themes:
                    campaign_data['visual_theme'] = theme
                    return campaign_data
                else:
                    return {
                        'error': f'Please choose from: {", ".join(valid_themes)}',
                        'suggestions': valid_themes
                    }
                
            elif step == 'call_to_action':
                cta = user_input.strip()
                valid_ctas = ['Shop Now', 'Learn More', 'Sign Up', 'Get Started', 
                              'Download Now', 'Book Now', 'Register Today', 'Join Us',
                              'Follow Us', 'Contact Us', 'Visit Website', 'Get Offer']
                
                if cta:
                    campaign_data['call_to_action'] = cta
                    return campaign_data
                else:
                    return {
                        'error': f'Please provide a clear call to action. Examples: {", ".join(valid_ctas)}',
                        'suggestions': valid_ctas
                    }
                
            elif step == 'additional_details':
                campaign_data['special_offers'] = user_input.strip()
                return campaign_data
                
            else:
                return campaign_data
                
        except Exception as e:
            app_logger.error(f"Error processing user input for step {step}: {e}")
            return campaign_data

    def generate_smart_suggestions(self, campaign_data: Dict[str, Any]) -> Dict[str, Any]:
        """Generate AI-powered suggestions based on campaign data"""
        
        try:
            campaign_type = campaign_data.get('type', 'sale')
            target_audience = campaign_data.get('target_audience', 'all_customers')
            platforms = campaign_data.get('platforms', ['facebook', 'instagram'])
            goal = campaign_data.get('goal', 'sales')
            
            # Get type-specific configurations
            type_config = self.campaign_types.get(campaign_type, self.campaign_types['sale'])
            audience_config = self.target_audiences.get(target_audience, self.target_audiences['all_customers'])
            
            # Generate visual theme
            visual_theme = type_config['visual_theme']
            
            # Generate call to action
            call_to_action = type_config['call_to_action']
            
            # Generate suggested hashtags
            base_hashtags = type_config['default_hashtags'].copy()
            custom_hashtags = self._generate_custom_hashtags(campaign_data)
            suggested_hashtags = list(set(base_hashtags + custom_hashtags))
            
            # Generate optimal posting times
            optimal_posting_times = audience_config['optimal_times']
            
            # Generate ad copy variations
            ad_copy_variations = []
            for template in type_config['ad_copy_templates']:
                ad_copy = template.format(
                    name=campaign_data.get('name', 'Campaign'),
                    special_offers=campaign_data.get('special_offers', 'Special Offer'),
                    cta=call_to_action
                )
                ad_copy_variations.append(ad_copy)
            
            # Generate platform-specific strategies
            platform_strategies = {}
            for platform in platforms:
                if platform in self.platform_strategies:
                    platform_strategies[platform] = self.platform_strategies[platform]
            
            # Generate poster URL
            poster_url = f"/posters/{campaign_data.get('name', 'campaign').lower().replace(' ', '_')}_2024.png"
            
            return {
                **campaign_data,
                'visual_theme': visual_theme,
                'call_to_action': call_to_action,
                'suggested_hashtags': suggested_hashtags,
                'optimal_posting_times': optimal_posting_times,
                'ad_copy_variations': ad_copy_variations,
                'platform_strategies': platform_strategies,
                'poster_url': poster_url,
                'content_focus': audience_config['content_focus']
            }
            
        except Exception as e:
            app_logger.error(f"Error generating smart suggestions: {e}")
            return campaign_data

    def create_campaign_from_chat(self, campaign_data: Dict[str, Any], user_token: str) -> Dict[str, Any]:
        """Create campaign from chatbot conversation data"""
        
        try:
            # Get user information
            current_user = get_user_from_token(user_token)
            if not current_user:
                raise Exception("Invalid user token")
            
            # Generate smart suggestions
            enhanced_data = self.generate_smart_suggestions(campaign_data)
            
            # Add required fields
            enhanced_data['created_by'] = current_user['username']
            enhanced_data['created_at'] = datetime.utcnow().isoformat()
            enhanced_data['companyid'] = current_user.get('companyid', 0)
            enhanced_data['status'] = 'draft'
            enhanced_data['leads'] = 0
            enhanced_data['conversion_rate'] = 0.0
            
            # Generate unique ID
            enhanced_data['id'] = self._generate_campaign_id()
            
            # Save to database
            if mongo_db.create_campaign(enhanced_data):
                # Generate poster (non-critical)
                try:
                    poster_url = poster_generator.generate_campaign_poster(enhanced_data)
                    enhanced_data['poster_url'] = poster_url
                    app_logger.info(f"Poster generated: {poster_url}")
                except Exception as e:
                    app_logger.warning(f"Failed to generate poster: {e}")
                    # Keep default poster URL
                    enhanced_data['poster_url'] = "/posters/default_campaign.png"
                
                app_logger.info(f"Campaign created via chatbot: {enhanced_data['name']}")
                return {
                    'success': True,
                    'campaign': enhanced_data,
                    'message': 'Campaign created successfully with AI assistance!'
                }
            else:
                raise Exception("Failed to save campaign to database")
                
        except Exception as e:
            app_logger.error(f"Error creating campaign from chat: {e}")
            return {
                'success': False,
                'error': str(e),
                'message': 'Failed to create campaign'
            }

    def _map_campaign_type(self, user_input: str) -> str:
        """Map user input to campaign type"""
        input_lower = user_input.lower()
        
        type_mapping = {
            'sale': ['sale', 'promotion', 'discount', 'offer', 'deal'],
            'product_launch': ['product launch', 'launch', 'new product', 'product'],
            'holiday': ['holiday', 'festive', 'christmas', 'thanksgiving', 'seasonal'],
            'brand_awareness': ['brand awareness', 'brand', 'awareness', 'identity'],
            'lead_generation': ['lead generation', 'leads', 'generation', 'business']
        }
        
        for campaign_type, keywords in type_mapping.items():
            if any(keyword in input_lower for keyword in keywords):
                return campaign_type
        
        return 'sale'  # Default

    def _map_target_audience(self, user_input: str) -> str:
        """Map user input to target audience"""
        input_lower = user_input.lower()
        
        audience_mapping = {
            'new_customers': ['new', 'potential', 'prospects'],
            'existing_customers': ['existing', 'current', 'loyal', 'repeat'],
            'all_customers': ['all', 'everyone', 'general'],
            'specific_demographics': ['specific', 'targeted', 'demographic', 'niche']
        }
        
        for audience, keywords in audience_mapping.items():
            if any(keyword in input_lower for keyword in keywords):
                return audience
        
        return 'all_customers'  # Default

    def _map_duration(self, user_input: str) -> int:
        """Map user input to duration in days"""
        input_lower = user_input.lower()
        
        if 'week' in input_lower:
            if '1' in input_lower or 'one' in input_lower:
                return 7
            elif '2' in input_lower or 'two' in input_lower:
                return 14
            elif '3' in input_lower or 'three' in input_lower:
                return 21
            elif '4' in input_lower or 'four' in input_lower:
                return 28
        elif 'month' in input_lower:
            if '1' in input_lower or 'one' in input_lower:
                return 30
            elif '2' in input_lower or 'two' in input_lower:
                return 60
        elif 'day' in input_lower:
            # Extract number of days
            import re
            numbers = re.findall(r'\d+', input_lower)
            if numbers:
                return int(numbers[0])
        
        return 14  # Default to 2 weeks

    def _map_platforms(self, user_input: str) -> List[str]:
        """Map user input to social media platforms"""
        input_lower = user_input.lower()
        platforms = []
        
        platform_mapping = {
            'facebook': ['facebook', 'fb'],
            'instagram': ['instagram', 'ig', 'insta'],
            'linkedin': ['linkedin', 'li'],
            'twitter': ['twitter', 'tweet']
        }
        
        for platform, keywords in platform_mapping.items():
            if any(keyword in input_lower for keyword in keywords):
                platforms.append(platform)
        
        # If no specific platforms mentioned, default to facebook & instagram
        if not platforms:
            platforms = ['facebook', 'instagram']
        
        return platforms

    def _map_goal(self, user_input: str) -> str:
        """Map user input to campaign goal"""
        input_lower = user_input.lower()
        
        goal_mapping = {
            'sales': ['sales', 'sell', 'revenue', 'purchase', 'buy'],
            'leads': ['leads', 'lead generation', 'prospects', 'sign up'],
            'engagement': ['engagement', 'engage', 'interaction', 'likes', 'comments'],
            'brand_awareness': ['brand awareness', 'awareness', 'brand', 'visibility']
        }
        
        for goal, keywords in goal_mapping.items():
            if any(keyword in input_lower for keyword in keywords):
                return goal
        
        return 'sales'  # Default

    def _generate_custom_hashtags(self, campaign_data: Dict[str, Any]) -> List[str]:
        """Generate custom hashtags based on campaign data"""
        hashtags = []
        
        # Add campaign name as hashtag
        name = campaign_data.get('name', '')
        if name:
            clean_name = name.replace(' ', '').replace('&', '').replace('!', '')
            hashtags.append(f"#{clean_name}")
        
        # Add type-specific hashtags
        campaign_type = campaign_data.get('type', '')
        if campaign_type:
            hashtags.append(f"#{campaign_type.replace('_', '').title()}")
        
        # Add year
        current_year = datetime.now().year
        hashtags.append(f"#{current_year}")
        
        return hashtags

    def _generate_campaign_id(self) -> int:
        """Generate unique campaign ID"""
        try:
            # Get existing campaigns to determine next ID
            existing_campaigns = mongo_db.get_campaigns()
            if existing_campaigns:
                max_id = max(campaign.get('id', 0) for campaign in existing_campaigns)
                return max_id + 1
            else:
                return 1
        except:
            return int(uuid.uuid4().hex[:8], 16) % 1000000

# Global instance
campaign_chatbot = CampaignChatbot()

File: src/components/ChatBot/CampaignChatBot.js
import React, { useState, useEffect, useRef } from 'react';
import {
    Box,
    Card,
    CardContent,
    Typography,
    TextField,
    Button,
    IconButton,
    Paper,
    Chip,
    Dialog,
    DialogTitle,
    DialogContent,
    DialogActions,
    Grid,
    CircularProgress,
    Alert
} from '@mui/material';
import {
    Send as SendIcon,
    Close as CloseIcon,
    SmartToy as BotIcon,
    Person as PersonIcon,
    AutoAwesome as SparkleIcon,
    Download as DownloadIcon
} from '@mui/icons-material';
import { useAuth } from '../../contexts/AuthContext';
import axios from 'axios';

const CampaignChatBot = ({ open, onClose, onCampaignCreated }) => {
    const { getAuthHeaders } = useAuth();
    const [messages, setMessages] = useState([]);
    const [inputValue, setInputValue] = useState('');
    const [isTyping, setIsTyping] = useState(false);
    const [currentStep, setCurrentStep] = useState('welcome');
    const [campaignData, setCampaignData] = useState({});
    const [showSummary, setShowSummary] = useState(false);
    const [posterPreview, setPosterPreview] = useState(null);
    const messagesEndRef = useRef(null);

    const scrollToBottom = () => {
        messagesEndRef.current?.scrollIntoView({ behavior: 'smooth' });
    };

    useEffect(() => {
        scrollToBottom();
    }, [messages]);

    useEffect(() => {
        if (open) {
            initializeChat();
        }
    }, [open]);

    const initializeChat = () => {
        const welcomeMessage = {
            id: Date.now(),
            type: 'bot',
            content: "👋 Hi! I'm your AI campaign assistant. I'll help you create an amazing marketing campaign! Let's start with some basic details.",
            timestamp: new Date().toISOString()
        };
        setMessages([welcomeMessage]);
        setCurrentStep('campaign_name');
        setCampaignData({});
        setShowSummary(false);
        setPosterPreview(null);
        
        setTimeout(() => {
            askQuestion("What would you like to name your campaign?");
        }, 1000);
    };

    const askQuestion = (question) => {
        const questionMessage = {
            id: Date.now(),
            type: 'bot',
            content: question,
            timestamp: new Date().toISOString()
        };
        setMessages(prev => [...prev, questionMessage]);
    };

    const addBotMessage = (content, suggestions = []) => {
        const message = {
            id: Date.now(),
            type: 'bot',
            content,
            suggestions,
            timestamp: new Date().toISOString()
        };
        setMessages(prev => [...prev, message]);
    };

    const addUserMessage = (content) => {
        const message = {
            id: Date.now(),
            type: 'user',
            content,
            timestamp: new Date().toISOString()
        };
        setMessages(prev => [...prev, message]);
    };

    const processUserInput = async (input) => {
        addUserMessage(input);
        setIsTyping(true);

        // Simulate processing delay
        await new Promise(resolve => setTimeout(resolve, 1000));

        switch (currentStep) {
            case 'campaign_name':
                setCampaignData(prev => ({ ...prev, name: input }));
                setCurrentStep('campaign_type');
                askQuestion("What type of campaign is this?", [
                    "Sale/Promotion",
                    "Product Launch", 
                    "Holiday Special",
                    "Brand Awareness",
                    "Lead Generation"
                ]);
                break;

            case 'campaign_type':
                const typeMap = {
                    'sale/promotion': 'sale',
                    'product launch': 'product_launch',
                    'holiday special': 'holiday',
                    'brand awareness': 'brand_awareness',
                    'lead generation': 'lead_generation'
                };
                const campaignType = typeMap[input.toLowerCase()] || 'sale';
                setCampaignData(prev => ({ ...prev, type: campaignType }));
                setCurrentStep('target_audience');
                askQuestion("Who are you targeting with this campaign?", [
                    "New Customers",
                    "Existing Customers",
                    "All Customers",
                    "Specific Demographics"
                ]);
                break;

            case 'target_audience':
                const audienceMap = {
                    'new customers': 'new_customers',
                    'existing customers': 'existing_customers',
                    'all customers': 'all_customers',
                    'specific demographics': 'specific_demographics'
                };
                const targetAudience = audienceMap[input.toLowerCase()] || 'all_customers';
                setCampaignData(prev => ({ ...prev, target_audience: targetAudience }));
                setCurrentStep('duration');
                askQuestion("How long will this campaign run?", [
                    "1 week",
                    "2 weeks",
                    "1 month",
                    "Custom duration"
                ]);
                break;

            case 'duration':
                const durationMap = {
                    '1 week': 7,
                    '2 weeks': 14,
                    '1 month': 30,
                    'custom duration': 21
                };
                const duration = durationMap[input.toLowerCase()] || 14;
                setCampaignData(prev => ({ ...prev, duration_days: duration }));
                setCurrentStep('budget');
                askQuestion("What's your budget range for this campaign?", [
                    "$500-1000",
                    "$1000-5000",
                    "$5000+",
                    "Limited budget"
                ]);
                break;

            case 'budget':
                setCampaignData(prev => ({ ...prev, budget_range: input }));
                setCurrentStep('platforms');
                askQuestion("Which social media platforms would you like to use?", [
                    "Facebook & Instagram",
                    "All Platforms",
                    "LinkedIn Only",
                    "Twitter Only"
                ]);
                break;

            case 'platforms':
                const platformMap = {
                    'facebook & instagram': ['facebook', 'instagram'],
                    'all platforms': ['facebook', 'instagram', 'linkedin', 'twitter'],
                    'linkedin only': ['linkedin'],
                    'twitter only': ['twitter']
                };
                const platforms = platformMap[input.toLowerCase()] || ['facebook', 'instagram'];
                setCampaignData(prev => ({ ...prev, platforms }));
                setCurrentStep('goal');
                askQuestion("What's the main goal of this campaign?", [
                    "Drive Sales",
                    "Generate Leads",
                    "Increase Engagement",
                    "Build Brand Awareness"
                ]);
                break;

            case 'goal':
                const goalMap = {
                    'drive sales': 'sales',
                    'generate leads': 'leads',
                    'increase engagement': 'engagement',
                    'build brand awareness': 'brand_awareness'
                };
                const goal = goalMap[input.toLowerCase()] || 'sales';
                setCampaignData(prev => ({ ...prev, goal }));
                setCurrentStep('visual_theme');
                askQuestion("What visual theme would you like for your poster?", [
                    "🌊 Blue Ocean (Professional & Clean)",
                    "💻 Modern Tech (Bold & Innovative)", 
                    "🎄 Festive Red (Celebratory & Bright)",
                    "🌿 Nature Green (Fresh & Organic)",
                    "👔 Elegant Purple (Sophisticated & Premium)",
                    "🌅 Sunset Orange (Warm & Energetic)",
                    "⚫ Elegant Black (Minimal & Professional)",
                    "📊 Professional Green (Corporate & Trustworthy)"
                ]);
                break;

            case 'visual_theme':
                const themeMap = {
                    'blue ocean': 'blue_ocean',
                    'modern tech': 'modern_tech',
                    'festive red': 'festive_red',
                    'nature green': 'nature_green',
                    'elegant purple': 'elegant_purple',
                    'sunset orange': 'sunset_orange',
                    'elegant black': 'elegant_black',
                    'professional green': 'professional_green'
                };
                const visualTheme = themeMap[input.toLowerCase()] || 'blue_ocean';
                setCampaignData(prev => ({ ...prev, visual_theme: visualTheme }));
                setCurrentStep('call_to_action');
                askQuestion("What should be the main call to action?", [
                    "Shop Now",
                    "Learn More", 
                    "Sign Up",
                    "Get Started",
                    "Download Now",
                    "Book Now",
                    "Register Today",
                    "Join Us",
                    "Contact Us",
                    "Visit Website",
                    "Get Offer"
                ]);
                break;

            case 'call_to_action':
                setCampaignData(prev => ({ ...prev, call_to_action: input }));
                setCurrentStep('additional_details');
                addBotMessage("Great! Now tell me more about your campaign. You can describe special offers, visual preferences, or any other details you'd like to include.", [
                    "20% OFF everything!",
                    "Buy 1 Get 1 Free",
                    "Limited Time - Don't Miss Out!",
                    "Free Shipping on Orders Over $50",
                    "Early Bird Special - First 50 Get 25% Off",
                    "Custom Offer"
                ], true);
                break;

            case 'additional_details':
                setCampaignData(prev => ({ ...prev, special_offers: input }));
                generateSmartSuggestions();
                break;

            default:
                break;
        }

        setIsTyping(false);
    };

    const generateSmartSuggestions = async () => {
        setIsTyping(true);
        
        // Simulate AI processing
        await new Promise(resolve => setTimeout(resolve, 2000));

        const suggestions = generateCampaignSuggestions(campaignData);
        setCampaignData(prev => ({ ...prev, ...suggestions }));

        addBotMessage("🎉 Based on your inputs, I've generated some smart suggestions for your campaign!");

        setTimeout(() => {
            addBotMessage(`📊 **Campaign Summary:**\n\n**Name:** ${campaignData.name}\n**Type:** ${campaignData.type}\n**Target:** ${campaignData.target_audience}\n**Duration:** ${campaignData.duration_days} days\n**Budget:** ${campaignData.budget_range}\n**Platforms:** ${campaignData.platforms?.join(', ')}\n**Goal:** ${campaignData.goal}\n**Special Offer:** ${campaignData.special_offers}\n\n🎨 **Visual Theme:** ${suggestions.visual_theme}\n📱 **Call to Action:** ${suggestions.call_to_action}\n🏷️ **Suggested Hashtags:** ${suggestions.suggested_hashtags?.join(', ')}\n⏰ **Best Posting Times:** ${suggestions.optimal_posting_times?.join(', ')}`);
            
            setTimeout(() => {
                addBotMessage("I'm also creating a professional poster for your campaign... 🎨");
                generatePosterPreview();
            }, 2000);
        }, 1000);

        setCurrentStep('review');
        setIsTyping(false);
    };

    const generateCampaignSuggestions = (data) => {
        const themes = {
            'sale': 'blue_ocean',
            'product_launch': 'modern_tech',
            'holiday': 'festive_red',
            'brand_awareness': 'elegant_black',
            'lead_generation': 'professional_green'
        };

        const callsToAction = {
            'sales': 'Shop Now',
            'leads': 'Sign Up',
            'engagement': 'Learn More',
            'brand_awareness': 'Discover'
        };

        const hashtags = {
            'sale': ['#Sale', '#SpecialOffer', '#Discount'],
            'product_launch': ['#NewProduct', '#Innovation', '#Launch'],
            'holiday': ['#Holiday', '#Festive', '#Special'],
            'brand_awareness': ['#Brand', '#Discover', '#Story'],
            'lead_generation': ['#Leads', '#Growth', '#Business']
        };

        return {
            visual_theme: themes[data.type] || 'blue_ocean',
            call_to_action: callsToAction[data.goal] || 'Learn More',
            suggested_hashtags: hashtags[data.type] || ['#Campaign', '#Marketing'],
            optimal_posting_times: ['9:00 AM', '6:00 PM'],
            ad_copy_variations: [
                `🔥 ${data.name} is here! ${data.special_offers}. ${callsToAction[data.goal]} →`,
                `✨ Exciting news about ${data.name}! Don't miss out. ${callsToAction[data.goal]}!`,
                `🎯 Amazing opportunity: ${data.name}. ${data.special_offers}. ${callsToAction[data.goal]} now!`
            ]
        };
    };

    const generatePosterPreview = () => {
        // Simulate poster generation
        setTimeout(() => {
            setPosterPreview(`/posters/${campaignData.name?.toLowerCase().replace(/\s+/g, '_')}_2024.png`);
            addBotMessage("🎨 Your professional campaign poster is ready! I'll now create your campaign with all these details.");
            setTimeout(() => {
                setShowSummary(true);
            }, 2000);
        }, 3000);
    };

    const handleSendMessage = () => {
        if (inputValue.trim()) {
            processUserInput(inputValue);
            setInputValue('');
        }
    };

    const handleSuggestionClick = (suggestion) => {
        setInputValue(suggestion);
    };

    const createCampaign = async () => {
        try {
            setIsTyping(true);
            const response = await axios.post(
                `${process.env.REACT_APP_API_LINKS}/api/v1/campaigns/chatbot-create`,
                campaignData,
                { headers: getAuthHeaders() }
            );

            addBotMessage("🎉 Congratulations! Your campaign has been successfully created! You can now view it in your campaigns dashboard.");
            
            setTimeout(() => {
                onCampaignCreated && onCampaignCreated(response.data.campaign);
                handleClose();
            }, 2000);

        } catch (error) {
            console.error('Error creating campaign:', error);
            
            // More detailed error handling
            let errorMessage = "❌ Sorry, I encountered an error creating your campaign. Please try again.";
            
            if (error.response) {
                // Server responded with error status
                if (error.response.status === 403) {
                    errorMessage = "❌ You don't have permission to create campaigns. Please contact your administrator.";
                } else if (error.response.status === 500) {
                    errorMessage = "⚠️ There was a server error, but your campaign might have been created. Please check your campaigns list.";
                } else if (error.response.data && error.response.data.detail) {
                    errorMessage = `❌ ${error.response.data.detail}`;
                }
            } else if (error.request) {
                // Network error
                errorMessage = "⚠️ Network error. Please check your connection and try again.";
            }
            
            addBotMessage(errorMessage);
        } finally {
            setIsTyping(false);
        }
    };

    const handleClose = () => {
        setMessages([]);
        setCampaignData({});
        setCurrentStep('welcome');
        setShowSummary(false);
        setPosterPreview(null);
        onClose();
    };

    const handleDownloadPoster = () => {
        if (posterPreview) {
            const link = document.createElement('a');
            link.href = posterPreview;
            link.download = `${campaignData.name?.replace(/\s+/g, '_') || 'campaign'}_poster.png`;
            document.body.appendChild(link);
            link.click();
            document.body.removeChild(link);
        }
    };

    const handleKeyPress = (e) => {
        if (e.key === 'Enter' && !e.shiftKey) {
            e.preventDefault();
            handleSendMessage();
        }
    };

    return (
        <Dialog open={open} onClose={handleClose} maxWidth="md" fullWidth>
            <DialogTitle>
                <Box display="flex" alignItems="center" justifyContent="space-between">
                    <Box display="flex" alignItems="center" gap={1}>
                        <SparkleIcon color="primary" />
                        <Typography variant="h6">AI Campaign Assistant</Typography>
                    </Box>
                    <IconButton onClick={handleClose}>
                        <CloseIcon />
                    </IconButton>
                </Box>
            </DialogTitle>
            
            <DialogContent>
                <Box sx={{ height: 500, display: 'flex', flexDirection: 'column' }}>
                    {/* Messages Area */}
                    <Box sx={{ flex: 1, overflow: 'auto', mb: 2, px: 1 }}>
                        {messages.map((message) => (
                            <Box key={message.id} sx={{ mb: 2 }}>
                                <Box display="flex" gap={1} alignItems="flex-start">
                                    {message.type === 'bot' ? (
                                        <BotIcon color="primary" sx={{ fontSize: 24 }} />
                                    ) : (
                                        <PersonIcon sx={{ fontSize: 24, color: 'grey.600' }} />
                                    )}
                                    <Box flex={1}>
                                        <Paper
                                            sx={{
                                                p: 2,
                                                bgcolor: message.type === 'bot' ? 'primary.50' : 'grey.100',
                                                borderRadius: 2
                                            }}
                                        >
                                            <Typography variant="body2" sx={{ whiteSpace: 'pre-line' }}>
                                                {message.content}
                                            </Typography>
                                        </Paper>
                                        {message.suggestions && (
                                            <Box sx={{ mt: 1, display: 'flex', flexWrap: 'wrap', gap: 0.5 }}>
                                                {message.suggestions.map((suggestion, index) => (
                                                    <Chip
                                                        key={index}
                                                        label={suggestion}
                                                        variant="outlined"
                                                        size="small"
                                                        clickable
                                                        onClick={() => handleSuggestionClick(suggestion)}
                                                        sx={{ cursor: 'pointer' }}
                                                    />
                                                ))}
                                            </Box>
                                        )}
                                    </Box>
                                </Box>
                            </Box>
                        ))}
                        
                        {isTyping && (
                            <Box display="flex" gap={1} alignItems="center">
                                <BotIcon color="primary" sx={{ fontSize: 24 }} />
                                <Paper sx={{ p: 2, bgcolor: 'primary.50', borderRadius: 2 }}>
                                    <CircularProgress size={16} />
                                    <Typography variant="body2" sx={{ ml: 1 }}>
                                        Thinking...
                                    </Typography>
                                </Paper>
                            </Box>
                        )}
                        
                        <div ref={messagesEndRef} />
                    </Box>

                    {/* Campaign Summary */}
                    {showSummary && campaignData && (
                        <Alert severity="success" sx={{ mb: 2 }}>
                            <Typography variant="subtitle2" gutterBottom>
                                📋 Campaign Ready to Create!
                            </Typography>
                            <Typography variant="body2">
                                <strong>{campaignData.name}</strong> • {campaignData.type} • {campaignData.duration_days} days
                            </Typography>
                            <Box sx={{ mt: 1 }}>
                                <Button
                                    variant="contained"
                                    size="small"
                                    onClick={createCampaign}
                                    disabled={isTyping}
                                >
                                    Create Campaign
                                </Button>
                            </Box>
                        </Alert>
                    )}

                    {/* Poster Preview */}
                    {posterPreview && (
                        <Box sx={{ mb: 2, textAlign: 'center' }}>
                            <Typography variant="subtitle2" gutterBottom>
                                🎨 Generated Poster Preview
                            </Typography>
                            <Paper sx={{ p: 2, bgcolor: 'grey.100' }}>
                                <img
                                    src={posterPreview}
                                    alt="Campaign poster preview"
                                    style={{
                                        maxWidth: '100%',
                                        height: 'auto',
                                        maxHeight: '200px',
                                        borderRadius: '8px'
                                    }}
                                    onError={(e) => {
                                        e.target.src = '/posters/default_campaign.png';
                                    }}
                                />
                                <Box mt={1} display="flex" gap={1} justifyContent="center">
                                    <Button
                                        size="small"
                                        variant="outlined"
                                        startIcon={<DownloadIcon />}
                                        onClick={handleDownloadPoster}
                                    >
                                        Download
                                    </Button>
                                </Box>
                                <Box mt={1}>
                                    <Typography variant="caption" color="textSecondary">
                                        {posterPreview}
                                    </Typography>
                                </Box>
                            </Paper>
                        </Box>
                    )}

                    {/* Input Area */}
                    <Box sx={{ borderTop: 1, borderColor: 'divider', pt: 2 }}>
                        <Box display="flex" gap={1}>
                            <TextField
                                fullWidth
                                variant="outlined"
                                placeholder="Type your response..."
                                value={inputValue}
                                onChange={(e) => setInputValue(e.target.value)}
                                onKeyPress={handleKeyPress}
                                disabled={isTyping || showSummary}
                                size="small"
                            />
                            <IconButton
                                color="primary"
                                onClick={handleSendMessage}
                                disabled={!inputValue.trim() || isTyping || showSummary}
                            >
                                <SendIcon />
                            </IconButton>
                        </Box>
                    </Box>
                </Box>
            </DialogContent>
        </Dialog>
    );
};

export default CampaignChatBot;
